Feature: Dynamic ice server add

vikramdattu opened the pull request:
## What changed
Added `peerConnectionUpdateIceServers()` public API to dynamically add ICE servers to an existing peer connection after creation. Internally implemented via `iceAgentAddIceServers()` in the ICE agent.

## Why
Allows applications to start peer connections immediately with just a STUN server, without waiting for TURN server configuration to be fetched. Once TURN servers become available, they can be added dynamically via the new API. This reduces peer connection establishment latency in most cases where local + STUN candidates suffice.

## How
- **Include.h**: Added `peerConnectionUpdateIceServers(PRtcPeerConnection, PRtcIceServer, UINT32)` public API declaration
- **IceAgent.c/h**: Implemented `iceAgentAddIceServers()` which validates inputs, checks for duplicates, parses new ICE servers, and initializes relay candidates for TURN servers (supporting both IPv4/IPv6, UDP/TCP)
- **PeerConnection.c**: Wrapper that delegates to the ICE agent
- **samples/common/Common.c**: Demonstrates deferred TURN server fetching — ICE config is fetched after peer connection creation in `signalingMessageReceived()` instead of during `initializePeerConnection()`

## Testing done
- Added 3 unit tests in `PeerConnectionApiTest.cpp`:
  - `peerConnectionUpdateIceServersNullArgs` — validates NULL/invalid argument handling
  - `peerConnectionUpdateIceServersSuccess` — validates successful STUN server addition
  - `peerConnectionUpdateIceServersDuplicate` — validates duplicate server handling
- All `PeerConnectionApiTest.*` tests pass (11/11)
- Clang-format compliance check passes
- Rebased onto latest `develop` branch (v1.17.0+)

vikramdattu committed:
Added support for dynamically adding ICE servers

 - With this change, the application can start peer connection with single ICE server
 - This will allow peer connection to run IceAgent instantaneously without waiting for TURN servers
 - Once, the TURN servers available, application can call `peerConnectionUpdateIceServers` to update it

In most of the cases, peer connection will take place with local + STUN candidates immediately

sirknightj commented on src/include/com/amazonaws/kinesis/video/webrtcclient/Include.h:
```diff
@@ -1848,6 +1848,21 @@ PUBLIC_API STATUS restartIce(PRtcPeerConnection);
  */
 PUBLIC_API STATUS closePeerConnection(PRtcPeerConnection);
 
+/**
```
Allows for pre-fetching of the host + stun candidates

sirknightj replied:
Wondering if you can modify the sample to use the new function? (testing the code path)

vikramdattu replied:
@sirknightj I'll modify the sample and check for if a test case can be added.

vikramdattu replied:
Test case added to showcase the behaviour.

YPandas committed:
fix(signaling): invalidate endpoint cache when storage is enabled (#2241)

* Added note that CMake 3.x is required and 4.x is not yet supported (#2234)

* fix(signaling): invalidate endpoint cache when storage is enabled

- Invalidate getEndpointTime when storageStatus becomes ENABLED
- Forces GetChannelEndpoint to request WEBRTC protocol endpoints
- Previously cached endpoints lacked WEBRTC (cached before storage enabled)
- Caused joinStorageSession() to fail with missing channelEndpointWebrtc
- Removed unnecessary caching logic in describeMediaStorageConf()
- Added explanatory comments about cache invalidation requirements

* feat: Add dynamic endpoint cache invalidation for storage status changes

Implement SIGNALING_API_CALL_CACHE_TYPE_FILE_EXCEPT_DESCRIBE_MEDIA caching
policy to optimize endpoint retrieval based on media storage configuration
changes when developer needs to change configuration frequently for testing or other purpose.

Key changes for the flag:
- Always call DescribeMediaStorageConfiguration API (bypass cache) to get
  real-time storage status
- Compare newly retrieved storage status with previously cached status
- Invalidate endpoint cache (getEndpointTime) when storage status changes
  between ENABLED and DISABLED to force fresh endpoint retrieval with
  correct protocol list (WEBRTC protocol only needed when storage enabled)
- Respect cache expiration period: fetch new endpoints if cache has expired
  even without storage status change
- Use cached endpoints only when both conditions met: cache not expired AND
  storage status unchanged

* Add .kiro and SignalingCache to gitignore (#2246)

* test: Add cache invalidation test for media storage toggle

- Run P2P master sample with storage disabled
- Enable storage, run storage audio/video master sample
- Disable storage, run P2P master sample again
- Verify all three runs succeed on the same channel
- Validates endpoint cache invalidation on storage status change
- fixed typo in Include.h

* test: Add cache invalidation tests for media storage toggle

- Add P2P → Storage → P2P test (cache-media-storage-test.yml)
- Add Storage → P2P → Storage test (cache-media-storage-reverse-test.yml)
- Verify endpoint cache invalidation on storage status change
- Provision and cleanup AWS resources per run

* fix: make sure the format is correct

* feat: add stuns support (#2243)

* Add stuns: URI parsing and validation

* Implement dtls-over-udp transport for stuns srflx gathering

* Use secure STUN sample endpoints for gov and adc regions

* Use FIPS endpoint naming convention in Gov/ADC regions

* Add documentation for early STUN DNS resolution and ICE secure transport paths

* Harden stuns: srflx candidate and ice server access checks

* Align ICE URI sizing and consolidate STUNS srflx invalidation

* Improve ICE server parsing diagnostics and srflx validation

* chore: fix clang lint

* Add integration test for srflx gathering using stuns:

* Avoid lock inversion when tearing down STUNS DTLS sessions

* Guard secure socket cleanup on successful lock acquisition

* version bump from 1.17.0 to 1.18.0 (#2250)

Co-authored-by: Yuqi Huang <yuuqih@amazon.com>

* ci: add JoinStorageSession log verification and set verbose logging in cache tests

- Set AWS_KVS_LOG_LEVEL to 1 (VERBOSE) for better CI debuggability
- Verify "Offer Received from JoinStorageSession Call" appears in logs

* chore: use FILE_EXCEPT_DESCRIBE_MEDIA caching policy in samples and document caching options

- Change default cachingPolicy in Common.c from SIGNALING_API_CALL_CACHE_TYPE_FILE
  to SIGNALING_API_CALL_CACHE_TYPE_FILE_EXCEPT_DESCRIBE_MEDIA so all samples
  automatically invalidate the endpoint cache when storage status changes
- Add caching policy documentation to README explaining all four policies

* ci: add cache file verification to media storage toggle tests

- Dump cache file contents before and after each storage status transition
  to make cache invalidation behavior visible in CI logs
- Verify JoinStorageSession succeeds after storage status changes, which
  proves FILE_EXCEPT_DESCRIBE_MEDIA correctly detected the stale cache
  and re-fetched endpoints with the right protocol list

* minor update to readme.md

---------

Co-authored-by: JC Yan <46540636+Ahhorse1@users.noreply.github.com>
Co-authored-by: Yuqi Huang <yuuqih@amazon.com>
Co-authored-by: Jeremy Gunawan <jggunawa@amazon.com>
Co-authored-by: jgooge <123762333+jgooge@users.noreply.github.com>

vikramdattu committed:
test: add unit tests for peerConnectionUpdateIceServers API

- peerConnectionUpdateIceServersNullArgs: validates NULL peer connection,
  NULL ice servers, and zero count return correct error codes
- peerConnectionUpdateIceServersSuccess: validates adding a STUN server
  to an existing peer connection succeeds
- peerConnectionUpdateIceServersDuplicate: validates duplicate servers
  are silently skipped while new ones are added

sirknightj commented on tst/PeerConnectionApiTest.cpp:
```diff
@@ -197,6 +197,70 @@ TEST_F(PeerConnectionApiTest, connectionState)
+    freePeerConnection(&pRtcPeerConnection);
+}
+
+TEST_F(PeerConnectionApiTest, peerConnectionUpdateIceServersDuplicate)
```
One suggestion for a test would be to create 2 peerconnections with `RtcConfiguration.iceTransportPolicy` = `ICE_TRANSPORT_POLICY_RELAY` and then use the new API to add the TURN servers async after some duration to verify the candidate gathering callback (new candidate generated, ready to send to other peer) and PeerConnection establishment

Should be able to build off the existing tests in `PeerConnectionFunctionalityTest.cpp`.

vikramdattu replied:
Good suggestion! I'll add a functional test in PeerConnectionFunctionalityTest.cpp as requested.

vikramdattu replied:
Added functional test `connectTwoPeersForcedTURNWithDynamicIceServers` — creates relay-only PCs, dynamically adds TURN servers, verifies connection establishes.

sirknightj commented on src/source/Ice/IceAgent.h:
```diff
@@ -301,6 +301,17 @@ STATUS createIceAgent(PCHAR, PCHAR, PIceAgentCallbacks, PRtcConfiguration, TIMER
 STATUS freeIceAgent(PIceAgent*);
 
+/**
+ * Add new ICE servers to an existing ICE agent and initialize relay candidates for TURN servers
```
What would happen if this was called after the ice agent timed out (no candidate pair within 10s?)

vikramdattu replied:
Currently, calling iceAgentAddIceServers after ICE agent timeout would succeed (the function only checks shutdown flag, which is set during explicit  shutdown, not timeout).
The new servers would be parsed and relay candidates initialized, but no connectivity checks would run since the state machine is already in FAILED state. It's benign though.

vikramdattu replied:
Should I add a check for the ICE agent state? If it's in FAILED state, we can return STATUS_INVALID_OPERATION to inform the caller.

vikramdattu replied:
Added state check — returns STATUS_INVALID_OPERATION if ICE agent is in FAILED state. Also added test `updateIceServersAfterIceFailure` to verify this behavior.

Ahhorse1 replied:
Another case is when this is called after ice agent gathering has finished, in which case these ICE servers would never checked. Perhaps a check can also be added for `pIceAgent->candidateGatheringFinished`

vikramdattu committed:
test: add functional test for dynamic ICE server addition with relay

Adds connectTwoPeersForcedTURNWithDynamicIceServers test that:
- Creates 2 PCs with ICE_TRANSPORT_POLICY_RELAY and only STUN server
- Dynamically adds TURN servers via peerConnectionUpdateIceServers
- Verifies relay candidates are generated and connection establishes

This test requires AWS credentials and network access (integration test).

vikramdattu committed:
Reject dynamic ICE server addition when agent is in FAILED state

Return STATUS_INVALID_OPERATION if iceAgentAddIceServers is called
after the ICE agent has transitioned to FAILED state, since new
candidates would not lead to connectivity checks at that point.

vikramdattu committed:
test: add test for ICE server update rejection after agent failure

Adds updateIceServersAfterIceFailure test that:
- Creates 2 relay-only PCs with no TURN servers
- Attempts connection which fails (no relay candidates)
- Verifies peerConnectionUpdateIceServers returns STATUS_INVALID_OPERATION
  after ICE agent has transitioned to FAILED state

vikramdattu committed:
fixup: format and adapt to develop API changes

- Update iceAgentInitRelayCandidate calls to 4-arg signature (added KVS_IP_FAMILY_TYPE)
- Add IPv4/IPv6 env var checks matching iceAgentInitRelayCandidates pattern
- Fix C99 mid-block variable declarations
- Apply clang-format

sirknightj commented on tst/PeerConnectionFunctionalityTest.cpp:
```diff
@@ -1280,6 +1280,116 @@ TEST_F(PeerConnectionFunctionalityTest, aggressiveNominationDTLSRaceConditionChe
+
+    deinitializeSignalingClient();
+}
+
```
Can you add this extra test for additional continuous verification?

This test verifies the new API after a sleep duration (closer pattern to expected use case). Passed on my local

```c
// Verify that dynamically adding TURN servers from a background thread
// 5 seconds after connectTwoPeers is called allows the connection to succeed
TEST_F(PeerConnectionFunctionalityTest, connectTwoPeersForcedTURNWithDelayedDynamicIceServers)
{
    ASSERT_EQ(TRUE, mAccessKeyIdSet);

    RtcConfiguration offerConfig, answerConfig;
    PRtcPeerConnection offerPc = NULL, answerPc = NULL;
    UINT32 i, uriCount = 0, iceConfigCount;
    PIceConfigInfo pIceConfigInfo;
    RtcIceServer turnServers[MAX_ICE_SERVERS_COUNT];
    UINT32 addTurnServerDelaySec = 5;

    ASSERT_TRUE(addTurnServerDelaySec < KVS_ICE_GATHER_REFLEXIVE_AND_RELAYED_CANDIDATE_TIMEOUT);

    initializeSignalingClient();

    // Use ALL policy on one side to have the ICE agent finish gathering Host+STUN candidates
    // This should validate that additional gathering can still be done after the fact
    MEMSET(&offerConfig, 0x00, SIZEOF(RtcConfiguration));
    offerConfig.iceTransportPolicy = ICE_TRANSPORT_POLICY_ALL;
    SNPRINTF(offerConfig.iceServers[0].urls, MAX_ICE_CONFIG_URI_LEN, KINESIS_VIDEO_STUN_URL, TEST_DEFAULT_REGION, TEST_DEFAULT_STUN_URL_POSTFIX);

    // Create PC with relay-only policy to verify TURN connection is used
    MEMSET(&answerConfig, 0x00, SIZEOF(RtcConfiguration));
    answerConfig.iceTransportPolicy = ICE_TRANSPORT_POLICY_RELAY;
    SNPRINTF(answerConfig.iceServers[0].urls, MAX_ICE_CONFIG_URI_LEN, KINESIS_VIDEO_STUN_URL, TEST_DEFAULT_REGION, TEST_DEFAULT_STUN_URL_POSTFIX);

    EXPECT_EQ(STATUS_SUCCESS, createPeerConnection(&offerConfig, &offerPc));
    EXPECT_EQ(STATUS_SUCCESS, createPeerConnection(&answerConfig, &answerPc));

    // Extract TURN servers from signaling
    MEMSET(turnServers, 0x00, SIZEOF(turnServers));
    ASSERT_EQ(STATUS_SUCCESS, signalingClientGetIceConfigInfoCount(mSignalingClientHandle, &iceConfigCount));
    for (uriCount = 0, i = 0; i < iceConfigCount; i++) {
        ASSERT_EQ(STATUS_SUCCESS, signalingClientGetIceConfigInfo(mSignalingClientHandle, i, &pIceConfigInfo));
        for (UINT32 j = 0; j < pIceConfigInfo->uriCount; j++) {
            STRNCPY(turnServers[uriCount].urls, pIceConfigInfo->uris[j], MAX_ICE_CONFIG_URI_LEN);
            STRNCPY(turnServers[uriCount].credential, pIceConfigInfo->password, MAX_ICE_CONFIG_CREDENTIAL_LEN);
            STRNCPY(turnServers[uriCount].username, pIceConfigInfo->userName, MAX_ICE_CONFIG_USER_NAME_LEN);
            uriCount++;
        }
    }
    ASSERT_TRUE(uriCount > 0);

    // Launch background thread that waits 5 seconds then adds TURN servers
    std::thread updateThread([offerPc, answerPc, &turnServers, uriCount, addTurnServerDelaySec]() {
        DLOGI("Wait %ds to add TURN servers", addTurnServerDelaySec);
        THREAD_SLEEP(addTurnServerDelaySec * HUNDREDS_OF_NANOS_IN_A_SECOND);
        DLOGI("Adding TURN servers now");
        EXPECT_EQ(STATUS_SUCCESS, peerConnectionUpdateIceServers(offerPc, turnServers, uriCount));
        EXPECT_EQ(STATUS_SUCCESS, peerConnectionUpdateIceServers(answerPc, turnServers, uriCount));
    });

    // connectTwoPeers polls for up to 100s, TURN servers arrive after 5s
    EXPECT_EQ(TRUE, connectTwoPeers(offerPc, answerPc));

    updateThread.join();

    closePeerConnection(offerPc);
    closePeerConnection(answerPc);

    freePeerConnection(&offerPc);
    freePeerConnection(&answerPc);

    deinitializeSignalingClient();
}
```

vikramdattu replied:
@sirknightj the test has been added in the latest commit. Tested locally to be passing as expected.

vikramdattu committed:
test: add connectTwoPeersForcedTURNWithDelayedDynamicIceServers (per PR #2164 review)

sirknightj commented:
Hi @vikramdattu, it seems like the [thread sanitizer is reporting a data race](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/actions/runs/25327513819/job/74427965760?pr=2164), could you take a look?

vikramdattu committed:
fix(ice): read iceAgentState under pIceAgent->lock in iceAgentAddIceServers

ThreadSanitizer surfaced a data race on PIceAgent::iceAgentState in the
new connectTwoPeersForcedTURNWithDelayedDynamicIceServers test:

* Reader: iceAgentAddIceServers() checked
  pIceAgent->iceAgentState != ICE_AGENT_STATE_FAILED before taking
  pIceAgent->lock.
* Writer: state-machine executors mutate iceAgentState from the timer
  thread under pIceAgent->lock (held by stepIceAgentStateMachine which
  wraps stepStateMachine).

Move the FAILED-state gate below MUTEX_LOCK so the read is also
synchronized via pIceAgent->lock. The shutdown flag check stays
outside the lock - it's already an atomic via ATOMIC_LOAD_BOOL.

vikramdattu commented:
Re https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/2164#issuecomment-4380859640 — pushed a small fix in 4cb230417e:

`iceAgentAddIceServers` was reading `pIceAgent->iceAgentState != ICE_AGENT_STATE_FAILED` before taking `pIceAgent->lock`. The state-machine executors that mutate `iceAgentState` already run under that lock (`stepIceAgentStateMachine` wraps `stepStateMachine`). Moved the FAILED-state gate below `MUTEX_LOCK` so the read is also synchronized. The `shutdown` flag check stays outside the lock — already an atomic via `ATOMIC_LOAD_BOOL`.

Verified locally with TSan: `connectTwoPeersForcedTURNWithDelayedDynamicIceServers` passes (no `WARNING: ThreadSanitizer`), and the touched-function tests (`IceAgentIceAgentAddIceServerUnitTest`, `peerConnectionUpdateIceServers{NullArgs,Success,Duplicate}`, `updateIceServersAfterIceFailure`, `connectTwoPeersForcedTURNWith{,Delayed}DynamicIceServers`) all pass.

I was not able to reproduce the issue on my side (5 pre-fix runs on macOS arm64 all passed, likely a Linux scheduling/timing window TSan only hits on the CI container). Re-running the CI would help confirm.

@sirknightj could you re-run CI please?

sirknightj committed:
Merge branch 'develop' into feature/dynamic_ice_server_add

unicornss reviewed (approved):
Thanks for the contribution - in the future it would be nice to have the sample demonstrating the usage and performance gain with this approach.

Ahhorse1 commented on src/source/Ice/IceAgent.c:
```diff
@@ -322,6 +322,128 @@ STATUS freeIceAgent(PIceAgent* ppIceAgent)
+
+    DLOGI("Adding %u new ICE servers to existing agent", iceServersCount);
+
+    MUTEX_LOCK(pIceAgent->lock);
```
Would recommend adding another shutdown check after obtaining the lock

Ahhorse1 reviewed (commented):
LGTM, left two minor comments for dicussion/future follow up
